src/browser/transports/injection_bus.ts

```diff
@@ -88,28 +91,84 @@ export default class NativeWindowInjectionBus extends EventEmitter {
         return;
       }
 
-      // Ack / Nack
+      this._listenerHeartbeat(parsedMessage);
+
       if (this._pendingRequests.has(messageId)) {
-        const isNack = parsedMessage.action.includes('error');
-        const pendingRequest = this._pendingRequests.get(messageId);
+        this._listenerAck(parsedMessage);
+      } else {
+        this._listenerBroadcastMsg(parsedMessage);
+      }
+    };
 
-        isNack
-          ? pendingRequest.reject((<NackMessage>parsedMessage).payload.reason)
-          : pendingRequest.resolve();
+    copyDataTransport.on('message', this._messageListener);
+
+    this.subscribe();
+  }
 
-        this._senderId = senderId;
-        this._pendingRequests.delete(messageId);
+  // Setup heartbeat
+  private setupHeartbeat() {
+    this._connected = false;
+    this._previousDllSequence = -1;
 
-        return;
+    this._listenerHeartbeat = (message: MessageBase): void => {
+      const { sequence } = message;
+
+      if (this._previousDllSequence + 1 !== sequence) {
+        // At least one message has been missed
+        const msg = `[NWI] Missed message(s) from ${this._senderId}: `
+          + `previous sequence ${this._previousDllSequence}, `
+          + `current sequence: ${sequence}`;
+        writeToLog('info', msg);
+      }
+
+      this._previousDllSequence = sequence;
+    };
+
+    // Ping message window every second and update current connection status
+    setInterval(() => {
+      this.send({
+        action: 'status/ping/request',
+        payload: {}
+      }).then(() => {
+        if (!this._connected) {
+          // This indicates that the core re-connected,
+          // so we need to subscribe to events again
+          this.subscribe();
+        }
+        this._connected = true;
+      }).catch(() => {
+        this._connected = false;
+      });
+    }, 1000);
+  }
+
+  // Setup listener responsible for parsing acks
+  private setupListenerAck() {
+    this._listenerAck = (message: MessageBase): void => {
+      const { messageId, senderId } = message;
+      const isNack = message.action.includes('error');
```

Comment:
:neckbeard: perhaps make this a enum to be explicit and to able to handle diff types of errors w/o having to fish through strings.

Reply:
Hmmm.. not sure it would work well for this case because of [the protocol](https://hackmd.io/z1ZaT6WcQlKMjweiycEP_w?view#ltnamespacegtlttopicgterror) 🤷‍♂

Reply:
Doesn't that say Int already or am I looking at the wrong thing? `code | <Integer> | The error code`

Reply:
It decides on whether it's an ack by parsing a part of a string vs looking at the property 🤷‍♂